### DISCUSSION TITLE
InputRules do not seem to search for matches correctly

### DISCUSSION THREAD

**Discussion by brentmparker** [ueberdosis/tiptap#2944] [General Questions]
I'm trying to integrate the prosemirror-math repository with TipTap 2.
[Semi-working example (codesandbox)](https://codesandbox.io/s/tiptap-prosemirror-math-j3f01r?file=/src/math-plugin.ts:4723-5322)
Feel free to tinker with the example

I can get the math to display, and I can edit it if I provide an initial state. But the input rules aren't working. 

For inline math, I should be able to type $$ and immediately insert an math-inline node. I should also be able to wrap content (e.g. $2000$) and that will wrap the text in an math-inline node.

What I think is happening is in TipTap Core InputRule.ts around line 120:

```ts
function run(config: {
  editor: Editor,
  from: number,
  to: number,
  text: string,
  rules: InputRule[],
  plugin: Plugin,
}): boolean {
  const {
    editor,
    from,
    to,
    text,
    rules,
    plugin,
  } = config
  const { view } = editor

  if (view.composing) {
    return false
  }

  const $from = view.state.doc.resolve(from)

  if (
    // check for code node
    $from.parent.type.spec.code
    // check for code mark
    || !!($from.nodeBefore || $from.nodeAfter)?.marks.find(mark => mark.type.spec.code)
  ) {
    return false
  }

  let matched = false

  /*****************************************/
  /*      Offending line here I think      */
  /*****************************************/
  const textBefore = getTextContentFromNodes($from) + text

  rules.forEach(rule => {
    if (matched) {
      return
    }
...
```

What this seems to be doing is getting ALL of the text from the entire node, then throwing the character that was just typed on the end.

For example, say there's a paragraph node containing
'This is a sentence'

If I insert a $ before a (e.g. 'This is $a sentence') `textBefore` becomes 'This is a sentence$'. 

When I add a `find` function to the inputRule, that is what is passed in as `text`.

Is this done intentionally? It seems strange that `text` would not be inserted in the position where it is actually typed. 

What recommendations do you have to work around this?

Thanks a lot!

**Comment by svenadlung** [comment]
@brentmparker, we have some known issues with the InputRules. I think your issue is related to https://github.com/ueberdosis/tiptap/issues/2937

I created a PR yesterday. Feel free to try out the changes and let me know if this works how it should:
https://github.com/ueberdosis/tiptap/pull/2941

**Comment by brentmparker** [comment]
@svenadlung Hey, thanks for the quick reply. I'm not sure this is the correct fix, though. It seems like in your fix you are only getting the text before the current cursor position, but there are instances where that might not lead to the expected result. For example, if you go to the preview generated in that fix ([https://deploy-preview-2941--tiptap-embed.netlify.app/preview/Examples/Book](https://deploy-preview-2941--tiptap-embed.netlify.app/preview/Examples/Book)) and try to get the input rule to create a horizontal rule (type ---), a horizontal rule is entered.  

Now type --, go left one character, and type - again.  You will have typed ---, but the input rule will not be triggered because the matching will only match -- rather than ---.  

What I think should be done is get the entire text from the node, and insert the newly typed character in the correct position. Then perform the regex matching.

Perhaps something like this might work:

```typescript
function run(config: {
  editor: Editor,
  from: number,
  to: number,
  text: string,
  rules: InputRule[],
  plugin: Plugin,
}): boolean {
  const {
    editor,
    from,
    to,
    text,
    rules,
    plugin,
  } = config
  const { view } = editor

  if (view.composing) {
    return false
  }

  const $from = view.state.doc.resolve(from)

  if (
    // check for code node
    $from.parent.type.spec.code
    // check for code mark
    || !!($from.nodeBefore || $from.nodeAfter)?.marks.find(mark => mark.type.spec.code)
  ) {
    return false
  }

  let matched = false

  /*******************
  *  Changes below *
  * ******************/

  const nodeText= getTextContentFromNodes($from);
  const nodeContent= nodeText.substring(0, $from.pos) + text + nodeText.substring($from.pos);

  rules.forEach(rule => {
    if (matched) {
      return
    }

    const match = inputRuleMatcherHandler(nodeContent, rule.find)

    if (!match) {
      return
    }
...
```

Or something like that. I'm far from an expert in how Prosemirror works (which is why I switched from Prosemirror to Tiptap. The Tiptap API has been much nicer that pure Prosemirror so far). Maybe instead of `$from.pos` it should be `$from.parentOffset` or whatever gives you the location that the new text should be inserted.

**Comment by brentmparker** [comment]
I did a little more digging. Here is the code Prosemirror uses to execute inputRules:

```typescript
function run(view: EditorView, from: number, to: number, text: string, rules: readonly InputRule[], plugin: Plugin) {
  if (view.composing) return false
  let state = view.state, $from = state.doc.resolve(from)
  if ($from.parent.type.spec.code) return false
  let textBefore = $from.parent.textBetween(Math.max(0, $from.parentOffset - MAX_MATCH), $from.parentOffset,
                                            null, "\ufffc") + text
  for (let i = 0; i < rules.length; i++) {
    let match = rules[i].match.exec(textBefore)
    let tr = match && rules[i].handler(state, match, from - (match[0].length - text.length), to)
    if (!tr) continue
    view.dispatch(tr.setMeta(plugin, {transform: tr, from, to, text}))
    return true
  }
  return false
}
```

In Tiptap, to acquire `textBefore`, the code is using

```typescript
import { ResolvedPos } from 'prosemirror-model'

export const getTextContentFromNodes = ($from: ResolvedPos, maxMatch = 500) => {
  let textBefore = ''

  $from.parent.nodesBetween(
    Math.max(0, $from.parentOffset - maxMatch),
    $from.parentOffset,
    (node, pos, parent, index) => {
      textBefore += node.type.spec.toText?.({
        node, pos, parent, index,
      }) || node.textContent || '%leaf%'
    },
  )

  return textBefore
}
```

In your fix the code has been changed to

```typescript
import { ResolvedPos } from 'prosemirror-model'

export const getTextContentFromNodes = ($from: ResolvedPos, maxMatch = 500) => {
  let textBefore = ''

  $from.parent.nodesBetween(
    Math.max(0, $from.parentOffset - maxMatch),
    $from.parentOffset,
    (node, pos, parent, index) => {
      textBefore += node.type.spec.toText?.({
        node, pos, parent, index,
      }) || $from.nodeBefore?.text || '%leaf%'
    },
  )

  return textBefore
}
```

I'm assuming this ends up with the same result as the Prosemirror code (only selecting the text content that comes before the cursor) and that seems like a reasonable solution.  Does this break anything else in TipTap? Are there other methods that rely on `getTextContentFromNodes` that expects the entire text content rather than just what is before the cursor?  Is there a reason why you couldn't just use the same code Prosemirror uses to get `textBefore` without modifying `getTextContentFromNodes` to prevent unforseen side-effects?
